packages/react-data/src/fetch/ReactDataFetch.js

```diff
@@ -0,0 +1,108 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import type {Wakeable} from 'shared/ReactTypes';
+
+import {readCache} from 'react/unstable-cache';
+
+const Pending = 0;
+const Resolved = 1;
+const Rejected = 2;
+
+type PendingResult = {|
+  status: 0,
+  value: Wakeable,
+|};
+
+type ResolvedResult = {|
+  status: 1,
+  value: mixed,
+|};
+
+type RejectedResult = {|
+  status: 2,
+  value: mixed,
+|};
+
+type Result = PendingResult | ResolvedResult | RejectedResult;
+
+const fetchKey = {};
+
+function readResultMap(): Map<string, Result> {
+  const resources = readCache().resources;
+  let map = resources.get(fetchKey);
+  if (map === undefined) {
+    map = new Map();
+    resources.set(fetchKey, map);
+  }
+  return map;
+}
+
+// TODO: options, auth, etc.
+export function fetch(url: string): Object {
+  const map = readResultMap();
+  const entry = map.get(url);
+  if (entry === undefined) {
+    let resolve = () => {};
+    const wakeable: Wakeable = new Promise(r => {
+      // TODO: should this be a plain thenable instead?
+      resolve = r;
+    });
+    const result: Result = {
+      status: Pending,
+      value: wakeable,
+    };
+    map.set(url, result);
+    const xhr = new XMLHttpRequest();
+    xhr.onload = function() {
+      // TODO: handle status codes.
+      // TODO: something more flexible than JSON.parse.
+      if (result.status !== Pending) {
+        return;
+      }
+      let responseJSON;
+      try {
+        responseJSON = JSON.parse(xhr.response);
+      } catch (err) {
+        const rejectedResult = ((result: any): RejectedResult);
+        rejectedResult.status = Rejected;
+        rejectedResult.value = err;
+        return;
+      }
+      const resolvedResult = ((result: any): ResolvedResult);
+      resolvedResult.status = Resolved;
+      resolvedResult.value = responseJSON;
```

Comment:
The return value should a a Response object like the Fetch API.

Reply:
The tricky part to polyfilling this with XHR is that we don't know if you're going to read it as binary data or string. I.e. if we should set it to `responseType = 'arraybuffer'`.

Reply:
So you check status and JSON.parse in userland? Kind of annoying for the common case. It's convenient when a library does this for you.